src/app/CommandSender.h

```diff
@@ -53,96 +53,158 @@ namespace app {
+         * allocated CommandSender object.
+         *
+         * This function will:
+         *      - Always be called exactly *once* for a given CommandSender instance.
+         *      - Be called even in error circumstances.
+         *      - Only be called after a successful call to SendCommandRequest returns, if SendCommandRequest is used.
+         *      - Always be called before a successful return from SendGroupCommandRequest, if SendGroupCommandRequest is used.
+         *
+         * This function must be implemented to destroy the CommandSender object.
+         *
+         * @param[in] apCommandSender   The command sender object of the terminated invoke command transaction.
+         */
+        virtual void OnDone(CommandSender * apCommandSender) = 0;
+    };
+
+    // CommandSender::ExtendedCallback::OnResponse is public SDK API, so we cannot break
+    // source compatibility for it. To allow for additional values to be added at a future
+    // time without constantly changing the function's declaration parameter list, we are
+    // defining the struct AdditionalResponseData and adding that to the parameter list
+    // to allow for future extendability.
+    struct AdditionalResponseData
+    {
+        Optional<uint16_t> commandRef;
+    };
+
+    // CommandSender::ExtendedCallback::OnError is public SDK API, so we cannot break source
+    // compatibility for it. To allow for additional values to be added at a future time
+    // without constantly changing the function's declaration parameter list, we are
+    // defining the struct ErrorData and adding that to the parameter list
+    // to allow for future extendability.
+    struct ErrorData
+    {
+        /**
+         * The following errors will be delivered through chipError
+         *
+         * - CHIP_ERROR_TIMEOUT: A response was not received within the expected response timeout.
+         * - CHIP_ERROR_*TLV*: A malformed, non-compliant response was received from the server.
+         * - CHIP_ERROR encapsulating a StatusIB: If we got a non-path-specific
+         *   status response from the server.  In that case,
+         *   StatusIB::InitFromChipError can be used to extract the status.
+         * - CHIP_ERROR*: All other cases.
+         */
+        CHIP_ERROR chipError;
+    };
+
+    /**
+     * @brief Callback that is extendable for future features, starting with batch commands
+     *
+     * The two major differences between ExtendedCallback and Callback are:
+     * 1. Path-specific errors go to OnResponse instead of OnError
+     *       - Note: Non-path-specific errors still go to OnError.
+     * 2. Instead of having new parameters at the end of the arguments list, with defaults,
+     *    as functionality expands, a parameter whose type is defined in this header is used
+     *    as the argument to the callbacks
+     *
+     * To support batch commands client needs to be using ExtendedCallback
+     */
+    class ExtendedCallback
```

Comment:
The comment says "extendable", and this would be a better name:
```suggestion
    class ExtendableCallback
```
and various documentation to be fixed accordingly.